core/src/nucleus/ribosome/callback/validate_entry.rs

```diff
@@ -1,53 +1,97 @@
 extern crate serde_json;
+use super::links_utils;
 use context::Context;
 use holochain_core_types::{
     dna::wasm::DnaWasm,
-    entry::{entry_type::EntryType, Entry},
+    entry::{entry_type::EntryType, Entry, ToEntry},
     error::HolochainError,
+    link::link_add::LinkAddEntry,
     validation::ValidationData,
 };
-use holochain_wasm_utils::api_serialization::validation::EntryValidationArgs;
+use holochain_wasm_utils::api_serialization::validation::{
+    EntryValidationArgs, LinkValidationArgs,
+};
 use nucleus::{
-    ribosome::{
-        self,
-        callback::{get_dna, get_wasm, CallbackResult},
-    },
+    ribosome::{self, callback::CallbackResult},
     ZomeFnCall,
 };
 use std::sync::Arc;
 
 pub fn validate_entry(
     entry: Entry,
-    entry_type: EntryType,
     validation_data: ValidationData,
     context: Arc<Context>,
 ) -> Result<CallbackResult, HolochainError> {
-    match entry_type {
+    match entry.entry_type() {
         EntryType::App(app_entry_type) => Ok(validate_app_entry(
-            entry,
-            app_entry_type,
+            entry.clone(),
+            app_entry_type.clone(),
             validation_data,
             context,
         )?),
         EntryType::Dna => Ok(CallbackResult::Pass),
+        EntryType::LinkAdd => Ok(validate_link_entry(
+            entry.clone(),
+            validation_data,
+            context,
+        )?),
         _ => Ok(CallbackResult::NotImplemented),
     }
 }
 
+fn validate_link_entry(
+    entry: Entry,
+    validation_data: ValidationData,
+    context: Arc<Context>,
+) -> Result<CallbackResult, HolochainError> {
+    let link_add_entry = LinkAddEntry::from_entry(&entry);
+    let link = link_add_entry.link().clone();
+    let (base, target) = links_utils::get_link_entries(&link, &context)?;
+    let link_definition_path = links_utils::find_link_definition_in_dna(
+        &base.entry_type(),
+        link.tag(),
+        &target.entry_type(),
+        &context,
+    ).ok_or(HolochainError::NotImplemented)?;
```

Comment:
What I see here implied is that this becomes part of the "system level" validation of links which makes it impossible for an app developer to create ad-hoc links manually (as we have done in holochain-proto) that aren't defined `entry!`  

I'm actually fine with this as MVP and getting off the ground because its conceptually simple, but I want to note that it decreases the expressivity from the initial design.  I think we might want have to do something that allows a "manual/custom" override for defining a link type with validation that's not inferred from the `entry!` definition.  But we can add that later.